File: examples/list.cs
union list_t {
  null_t
  cons_t
} {
  t: type
}

class null_t {
  t: type
}

class cons_t {
  t: type
  car: t
  cdr: list_t (t)
}

list_length: (
  [implicit]: { t: type }
  list: list_t (t)
) -> nat_t = case (list) {
  null_t => zero_t ()
  cons_t => succ_t (list_length (list.cdr))
}

list_append: (
  [implicit]: { t: type }
  ante: list_t (t),
  succ: list_t (t),
) -> list_t (t) = case (ante) {
  null_t => succ
  cons_t => cons_t (ante.car, list_append (ante.cdr, succ))
}

list_map: (
  [implicit]: { a: type, b: type }
  fun: (a) -> b,
  list: list_t (a),
) -> list_t (b) = case (list) {
  null_t => list
  cons_t => cons_t (fun (list.car), list_map (fun, list.cdr))
}

list_remove_first: (
  [implicit]: { t: type }
  x: t,
  list: list_t (t),
) -> list_t (t) = case (list) {
  null_t => list
  cons_t => case (eq_p (list.car, x)) {
    true_t => list.cdr
    false_t => cons_t (list.car, list_remove_first (list.cdr, x))
  }
}

union list_length_t {
  zero_length_t
  succ_length_t
} {
  [implicit]: { t: type }
  list : list_t (t)
  length : nat_t
}

class zero_length_t {
  list: list_t (t) = null_t ()
  length: nat_t = zero_t ()
}

class succ_length_t {
  list: list_t (t) = cons_t (x, l)
  length: nat_t = succ_t (n)
  prev: list_length_t (l, n)
}

/** in prolog, we will have:
 * append([], Succ, Succ).
 * append([Car | Cdr], Succ, [Car | ResultCdr]):-
 *   append(Cdr, Succ, ResultCdr).
 */

union list_append_t {
  zero_append_t
  succ_append_t
} {
  [implicit]: { t: type }
  ante: list_t (t)
  succ: list_t (t)
  result: list_t (t)
}

class zero_append_t {
  [implicit]: { t: type }
  ante: list_t (t) = null_t ()
  succ: list_t (t)
  result: list_t (t) = succ
}

class succ_append_t {
  [implicit]: {
    t: type
    car: t
  }
  ante: cons_t (t, car)
  succ: list_t (t)
  result: cons_t (t, car)
  prev: list_append_t (
    ante.cdr,
    succ,
    result.cdr,
  )
}

File: src/main/scala/xieyuheng/cicada/with_logic_variable/prelude/list.scala
package xieyuheng.cicada.with_logic_variable.prelude

import xieyuheng.cicada.with_logic_variable._
import xieyuheng.cicada.with_logic_variable.expDSL._

object list extends Module {

  import_all(nat)

  define_type("list_t", %("A" -> the_type),
    members = %(
      "null_t" -> %(),
      "cons_t" -> %(
        "head" -> the("A"),
        "tail" -> the("list_t" ap %("A" -> "A")))))

  define_fn("cdr",
    args = %("list" -> the("list_t")),
    ret = the("list_t"),
    body = "list" dot "tail")

  define_fn("list_length",
    args = %("list" -> the("list_t")),
    ret = the("nat_t"),
    body = choice("list", %(
      "null_t" -> "zero_t",
      "cons_t" -> ("succ_t" ap %(
        "prev" -> ("list_length" ap %(
          "list" -> ("list" dot "tail"))))))))

  define_fn("list_append",
    args = %(
      "ante" -> the("list_t"),
      "succ" -> the("list_t")),
    ret = the("list_t"),
    body = choice("ante", %(
      "null_t" -> "succ",
      "cons_t" -> ("cons_t" ap %(
        "A" -> ("ante" dot "A"),
        "head" -> ("ante" dot "head"),
        "tail" -> ("list_append" ap %(
          "ante" -> ("ante" dot "tail"),
          "succ" -> "succ")))))))

  define_fn("list_map",
    args = %(
      "A" :: Type(),
      "B" :: Type(),
      "f" :: pi(%("x" :: "A"), the("B")),
      "list" :: ("list_t" ap %("A" := "A"))),
    ret = the("list_t" ap %("A" := "B")),
    body = choice("list", %(
      "null_t" := "null_t",
      "cons_t" := ("cons_t" ap %(
        "A" := "B",
        "head" := ("f" ap %("x" -> ("list" dot "head"))),
        "tail" := ("list_map" ap %(
          "A" := "A",
          "B" := "B",
          "f" := "f",
          "list" := ("list" dot "tail"))))))))

  // define_fn("list_map",
  //   args = %(
  //     "A" -> the_type,
  //     "B" -> the_type,
  //     "f" -> the(pi(%("x" -> the("A")), the("B"))),
  //     "list" -> the("list_t" ap %("A" -> "A"))),
  //   ret = the("list_t" ap %("A" -> "B")),
  //   body = choice("list", %(
  //     "null_t" -> "null_t",
  //     "cons_t" -> ("cons_t" ap %(
  //       "A" -> "B",
  //       "head" -> ("f" ap %("x" -> ("list" dot "head"))),
  //       "tail" -> ("list_map" ap %(
  //         "A" -> "A",
  //         "B" -> "B",
  //         "f" -> "f",
  //         "list" -> ("list" dot "tail"))))))))

}
